pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,417 @@
+				// but are failing to list from a previous replay.
+				for _, gvk := range gvksFailingTolist.Items() {
+					if c.gvkAggregator.IsPresent(gvk) {
+						c.gvksToRelist.Add(gvk)
+					}
+				}
+
+				// save all gvks that need relisting
+				gvksToRelistForLoop := c.gvksToRelist.Items()
+
+				// clean state
+				gvksFailingTolist = watch.NewSet()
+				c.gvksToRelist = watch.NewSet()
+
+				stopChan = make(chan bool)
+
+				go c.replayLoop(ctx, gvksToRelistForLoop, stopChan)
+				go gvksFailingToListReconciler(stopChan)
+			}
+			c.mu.Unlock()
+		}
+	}
+}
+
+func (c *CacheManager) replayLoop(ctx context.Context, gvksToRelist []schema.GroupVersionKind, stopChan <-chan bool) {
+	for _, gvk := range gvksToRelist {
+		select {
+		case <-ctx.Done():
+			return
+		case <-stopChan:
+			return
+		default:
+			backoff := wait.Backoff{
+				Duration: time.Second,
+				Factor:   2,
+				Jitter:   0.1,
+				Steps:    3,
+			}
+
+			operation := func() (bool, error) {
+				if err := c.listAndSyncDataForGVK(ctx, gvk); err != nil {
+					return false, err
+				}
+
+				return true, nil
+			}
+
+			if err := wait.ExponentialBackoff(backoff, operation); err != nil {
+				log.Error(err, "internal: error listings gvk cache data", "gvk", gvk)
+			}
+		}
+	}
+}
+
+// listAndSyncData returns a set of gvks that were successfully listed and synced.
+func (c *CacheManager) listAndSyncData(ctx context.Context, gvks []schema.GroupVersionKind) *watch.Set {
+	gvksSuccessfullySynced := watch.NewSet()
+	for _, gvk := range gvks {
+		err := c.listAndSyncDataForGVK(ctx, gvk)
+		if err != nil {
+			log.Error(err, "internal: error syncing gvks cache data")
+			// we don't remove this gvk as we will try to re-add it later
+			// we also don't return on this error to be able to list and sync
+			// other gvks in order to protect against a bad gvk.
+		} else {
+			gvksSuccessfullySynced.Add(gvk)
+		}
+	}
+	return gvksSuccessfullySynced
+}
+
+// makeUpdates performs a conditional wipe followed by a replay if necessary as
+// given by the current spec (currentGVKsInAgg, excluderChanged) at the time of the call.
+func (c *CacheManager) makeUpdates(ctx context.Context) {
+	// assumes the caller has lock
+
+	currentGVKsInAgg := watch.NewSet()
+	currentGVKsInAgg.Add(c.gvkAggregator.ListAllGVKs()...)
+
+	if c.watchedSet.Equals(currentGVKsInAgg) && !c.excluderChanged {
+		// nothing to do if both sets are the same and the excluder didn't change
+		// and there are no gvks that need relisting from a previous wipe
+		return
+	}
+
+	gvksToDelete := c.watchedSet.Difference(currentGVKsInAgg)
+	newGVKsToSync := currentGVKsInAgg.Difference(c.watchedSet)
+	gvksToReplay := c.watchedSet.Intersection(currentGVKsInAgg)
+
+	if gvksToDelete.Size() != 0 || newGVKsToSync.Size() != 0 {
+		// in this case we need to replace the watch set again since there
+		// is drift between the aggregator and the currently watched gvks
+		if err := c.replaceWatchSet(ctx, currentGVKsInAgg); err != nil {
```

Comment:
Since the watchSet is derived exactly from the Aggregator, we could just do this kind of work in `AddSources` and `RemoveSources`.  This is helpful because we won't have to worry about handling amounts of time where the `watchSet` and the `GVKAggregator` are different.

Reply:
That also limits the scope of what the background process has to do.  It needs only check if there is a reason to list and then list.

Reply:
Chatted offline, we think that adding the `replaceWatchSet` call as part of `RemoveSource` func will simplify the code and does not constitute a concern from a load perspective to be protected in the offline path.